app/src/main/java/org/septa/android/app/services/apiinterfaces/PushNotificationService.java

```diff
@@ -0,0 +1,141 @@
+                String token = instanceIdResult.getToken();
+                // Do whatever you want with your token now
+                // i.e. store it on SharedPreferences or DB
+                // or directly send it to server
+
+
+                //for now we are displaying the token in the log
+                //copy it as this method is called only when the new token is generated
+                //and usually new token is only generated when the app is reinstalled or the data is cleared
+                Log.d(TAG, "TOKEN: " + token);
+            }
+        });
+    }
+
+    @Override
+    public void onMessageReceived(RemoteMessage remoteMessage) {
+        super.onMessageReceived(remoteMessage);
+
+        Log.d(TAG, remoteMessage.toString());
+
+        // check notification subscription window
+        if (PushNotificationManager.isWithinNotificationWindow(getApplicationContext())) {
+
+            if (remoteMessage.getData().size() > 0) {
+                Map<String, String> data = remoteMessage.getData();
+
+                // get notification title / body
+                final NotificationType notificationType = NotificationType.valueOf(data.get(NOTIFICATION_KEY_TYPE));
+                final String message = data.get(NOTIFICATION_KEY_MESSAGE);
+
+                // send notification if special announcement
+                if (notificationType == NotificationType.SPECIAL_ANNOUNCEMENT) {
+                    PushNotificationManager.getInstance(getApplicationContext()).buildSpecialAnnouncementNotification(getApplicationContext(), message);
+
+                } else {
+                    // get route ID for route specific alerts
+                    final String routeId = data.get(NOTIFICATION_KEY_ROUTE_ID);
+
+                    // send notification if rail delay
+                    if (notificationType == NotificationType.DELAY) {
+                        // handle the data message
+                        final String vehicleId = data.get(NOTIFICATION_KEY_VEHICLE_ID);
+                        final String destinationStopId = data.get(NOTIFICATION_KEY_DESTINATION_STOP_ID);
+                        final DelayNotificationType delayType = DelayNotificationType.valueOf(data.get(NOTIFICATION_KEY_DELAY_TYPE));
+
+                        // default expiration date 3 hours from now
+                        Date expirationTimeStamp = addHoursToDate(new Date(), 3);
+
+                        // parse expiration date
+                        SimpleDateFormat df = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSSSSXXX", Locale.US);
+                        try {
+                            String expires = data.get(NOTIFICATION_KEY_EXPIRES);
+                            expirationTimeStamp = df.parse(expires);
+                            Log.d(TAG, "Parsing Expiration Time Stamp: " + expirationTimeStamp);
+                        } catch (ParseException e) {
+                            e.printStackTrace();
```

Comment:
Should get some crashlytics events here on the error handling.  Would be good to get if for some reason folks are missing notifications.